mbed-events/EventQueue.h

```diff
@@ -0,0 +1,764 @@
+     *
+     *  If called while the event queue's dispatch loop is active, the cancel
+     *  function does not garuntee that the event will not execute after it
+     *  returns, as the event may have already begun executing.
+     *
+     *  @param id       Unique id of the event
+     */
+    void cancel(int id);
+
+    /** Background an event queue onto a single-shot timer
+     *
+     *  The provided update function will be called to indicate when the queue
+     *  should be dispatched. A negative timeout will be passed to the update
+     *  function when the time is no longer needed.
+     *
+     *  Passing a null update function disables the existing timre.
+     *
+     *  The background function allows an event queue to take advantage of
+     *  hardware timers or even other event loops, allowing an event queue to
+     *  be effectively backgrounded.
+     *
+     *  @param update   Function called to indicate when the queue should be
+     *                  dispatched
+     */
+    void background(mbed::Callback<void(int)> update);
+
+    /** Chain an event queue onto another event queue
+     *
+     *  After chaining a queue to a target, calling dispatch on the target
+     *  queue will also dispatch events from this queue. The queues use
+     *  their own buffers and events must be handled independently.
+     *
+     *  A null queue as the target will unchain the existing queue.
+     *
+     *  The chain function allows multiple event queuest to be composed,
+     *  sharing the context of a dispatch loop while still being managed
+     *  independently
+     *
+     *  @param target   Queue that will dispatch this queue's events as a
+     *                  part of its dispatch loop
+     */
+    void chain(EventQueue *target);
+
+    /** Post an event to the queue
+     *
+     *  The specified callback will be executed in the context of the event
+     *  queue's dispatch loop.
+     *
+     *  The call function is irq safe and can act as a mechanism for moving
+     *  events out of irq contexts.
+     *
+     *  @param f        Function to execute in the context of the dispatch loop
+     *  @param a0..a4   Arguments to pass to the callback
+     *  @return         A unique id that represents the posted event and can
+     *                  be passed to cancel, or an id of 0 if there is not
+     *                  enough memory to allocate the event.
+     */
+    template <typename F>
+    int call(F f) {
+        struct local {
+            static void call(void *p) { (*static_cast<F*>(p))(); }
+            static void dtor(void *p) { static_cast<F*>(p)->~F(); }
+        };
+
+        void *p = equeue_alloc(&_equeue, sizeof(F));
+        if (!p) {
+            return 0;
+        }
+
+        F *e = new (p) F(f);
+        equeue_event_dtor(e, &local::dtor);
+        return equeue_post(&_equeue, &local::call, e);
+    }
+
+    template <typename F, typename A0>
+    int call(F f, A0 a0) {
```

Comment:
It would be better to have a documentation block for all the overloads, otherwise it will not be seen in the documentation.

`@see` directive  should be enough.